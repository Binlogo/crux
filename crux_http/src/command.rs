//! The Command based API for crux_http

use std::{fmt, future::Future, marker::PhantomData};

use crux_core::{command, Command};
use http_types::{
    convert::DeserializeOwned,
    headers::{HeaderName, ToHeaderValues},
    Body, Method, Mime, Url,
};
use serde::Serialize;

use crate::expect::{ExpectBytes, ExpectJson, ExpectString};
use crate::protocol::{HttpRequest, ProtocolRequestBuilder};
use crate::{expect::ResponseExpectation, protocol::HttpResult};
use crate::{middleware::Middleware, Response};
use crate::{HttpError, Request};

pub struct Http<Effect, Event> {
    effect: PhantomData<Effect>,
    event: PhantomData<Event>,
}

impl<Effect, Event> Http<Effect, Event>
where
    Effect: Send + From<crux_core::Request<HttpRequest>> + 'static,
    Event: Send + 'static,
{
    pub fn get(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Get, url.as_ref().parse().unwrap())
    }

    pub fn head(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Head, url.as_ref().parse().unwrap())
    }

    pub fn post(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Post, url.as_ref().parse().unwrap())
    }

    pub fn put(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Put, url.as_ref().parse().unwrap())
    }

    pub fn delete(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Delete, url.as_ref().parse().unwrap())
    }

    pub fn patch(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Patch, url.as_ref().parse().unwrap())
    }

    pub fn options(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Options, url.as_ref().parse().unwrap())
    }

    pub fn trace(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Trace, url.as_ref().parse().unwrap())
    }

    pub fn connect(url: impl AsRef<str>) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(Method::Connect, url.as_ref().parse().unwrap())
    }

    pub fn request(method: Method, url: Url) -> RequestBuilder<Effect, Event> {
        RequestBuilder::new(method, url)
    }
}

/// Request Builder
///
/// Provides an ergonomic way to chain the creation of a request.
/// This is generally accessed as the return value from `Http::{method}()`.
///
/// # Examples
///
/// ```no_run
/// use crux_http::http::{mime::HTML};
/// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
/// # struct Capabilities { http: crux_http::Http<Event> }
/// # fn update(caps: &Capabilities) {
/// caps.http
///     .post("https://httpbin.org/post")
///     .body("<html>hi</html>")
///     .header("custom-header", "value")
///     .content_type(HTML)
///     .send(Event::ReceiveResponse)
/// # }
/// ```
#[must_use]
pub struct RequestBuilder<Effect, Event, ExpectBody = Vec<u8>> {
    /// Holds the state of the request.
    req: Option<Request>,
    effect: PhantomData<Effect>,
    event: PhantomData<fn() -> Event>,
    expectation: Box<dyn ResponseExpectation<Body = ExpectBody> + Send>,
}

impl<Effect, Event> RequestBuilder<Effect, Event, Vec<u8>>
where
    Effect: Send + From<crux_core::Request<HttpRequest>> + 'static,
    Event: 'static,
{
    pub(crate) fn new(method: Method, url: Url) -> Self {
        Self {
            req: Some(Request::new(method, url)),
            effect: PhantomData,
            event: PhantomData,
            expectation: Box::new(ExpectBytes),
        }
    }
}

impl<Effect, Event, ExpectBody> RequestBuilder<Effect, Event, ExpectBody>
where
    Effect: Send + From<crux_core::Request<HttpRequest>> + 'static,
    Event: Send + 'static,
    ExpectBody: 'static,
{
    /// Sets a header on the request.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// caps.http
    ///     .get("https://httpbin.org/get")
    ///     .body("<html>hi</html>")
    ///     .header("header-name", "header-value")
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn header(mut self, key: impl Into<HeaderName>, value: impl ToHeaderValues) -> Self {
        self.req.as_mut().unwrap().insert_header(key, value);
        self
    }

    /// Sets the Content-Type header on the request.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # use crux_http::http::mime;
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// caps.http
    ///     .get("https://httpbin.org/get")
    ///     .content_type(mime::HTML)
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn content_type(mut self, content_type: impl Into<Mime>) -> Self {
        self.req
            .as_mut()
            .unwrap()
            .set_content_type(content_type.into());
        self
    }

    /// Sets the body of the request from any type with implements `Into<Body>`, for example, any type with is `AsyncRead`.
    /// # Mime
    ///
    /// The encoding is set to `application/octet-stream`.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// use serde_json::json;
    /// use crux_http::http::mime;
    /// caps.http
    ///     .post("https://httpbin.org/post")
    ///     .body(json!({"any": "Into<Body>"}))
    ///     .content_type(mime::HTML)
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn body(mut self, body: impl Into<Body>) -> Self {
        self.req.as_mut().unwrap().set_body(body);
        self
    }

    /// Pass JSON as the request body.
    ///
    /// # Mime
    ///
    /// The encoding is set to `application/json`.
    ///
    /// # Errors
    ///
    /// This method will return an error if the provided data could not be serialized to JSON.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # use serde::{Deserialize, Serialize};
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// #[derive(Deserialize, Serialize)]
    /// struct Ip {
    ///     ip: String
    /// }
    ///
    /// let data = &Ip { ip: "129.0.0.1".into() };
    /// caps.http
    ///     .post("https://httpbin.org/post")
    ///     .body_json(data)
    ///     .expect("could not serialize body")
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn body_json(self, json: &impl Serialize) -> crate::Result<Self> {
        Ok(self.body(Body::from_json(json)?))
    }

    /// Pass a string as the request body.
    ///
    /// # Mime
    ///
    /// The encoding is set to `text/plain; charset=utf-8`.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// caps.http
    ///     .post("https://httpbin.org/post")
    ///     .body_string("hello_world".to_string())
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn body_string(self, string: String) -> Self {
        self.body(Body::from_string(string))
    }

    /// Pass bytes as the request body.
    ///
    /// # Mime
    ///
    /// The encoding is set to `application/octet-stream`.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// caps.http
    ///     .post("https://httpbin.org/post")
    ///     .body_bytes(b"hello_world".to_owned())
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn body_bytes(self, bytes: impl AsRef<[u8]>) -> Self {
        self.body(Body::from(bytes.as_ref()))
    }

    /// Pass form data as the request body. The form data needs to be
    /// serializable to name-value pairs.
    ///
    /// # Mime
    ///
    /// The `content-type` is set to `application/x-www-form-urlencoded`.
    ///
    /// # Errors
    ///
    /// An error will be returned if the provided data cannot be serialized to
    /// form data.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # use std::collections::HashMap;
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// let form_data = HashMap::from([
    ///     ("name", "Alice"),
    ///     ("location", "UK"),
    /// ]);
    /// caps.http
    ///     .post("https://httpbin.org/post")
    ///     .body_form(&form_data)
    ///     .expect("could not serialize body")
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn body_form(self, form: &impl Serialize) -> crate::Result<Self> {
        Ok(self.body(Body::from_form(form)?))
    }

    /// Set the URL querystring.
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # use serde::{Deserialize, Serialize};
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    /// #[derive(Serialize, Deserialize)]
    /// struct Index {
    ///     page: u32
    /// }
    ///
    /// let query = Index { page: 2 };
    /// caps.http
    ///     .post("https://httpbin.org/post")
    ///     .query(&query)
    ///     .expect("could not serialize query string")
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn query(mut self, query: &impl Serialize) -> std::result::Result<Self, HttpError> {
        self.req.as_mut().unwrap().set_query(query)?;

        Ok(self)
    }

    /// Push middleware onto a per-request middleware stack.
    ///
    /// **Important**: Setting per-request middleware incurs extra allocations.
    /// Creating a `Client` with middleware is recommended.
    ///
    /// Client middleware is run before per-request middleware.
    ///
    /// See the [middleware] submodule for more information on middleware.
    ///
    /// [middleware]: ../middleware/index.html
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Vec<u8>>>) }
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// # fn update(caps: &Capabilities) {
    ///
    /// caps.http
    ///     .get("https://httpbin.org/redirect/2")
    ///     .middleware(crux_http::middleware::Redirect::default())
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn middleware(mut self, middleware: impl Middleware) -> Self {
        self.req.as_mut().unwrap().middleware(middleware);
        self
    }

    /// Return the constructed `Request` in a [`crux_core::command::RequestBuilder`].
    pub fn build(
        self,
    ) -> command::RequestBuilder<
        Effect,
        Event,
        impl Future<Output = Result<Response<ExpectBody>, HttpError>>,
    > {
        let req = self.req.expect("RequestBuilder::build called twice");

        command::RequestBuilder::new(|ctx| async move {
            let operation = req
                .into_protocol_request()
                .await
                .expect("should be able to convert request to protocol request");

            let result = Command::request_from_shell(operation)
                .into_future(ctx)
                .await;

            match result {
                HttpResult::Ok(response) => Response::<Vec<u8>>::new(response.into())
                    .await
                    .and_then(|r| self.expectation.decode(r)),
                HttpResult::Err(error) => Err(error),
            }
        })
    }

    /// Decode a String from the response body prior to dispatching it to the apps `update`
    /// function.
    ///
    /// This has no effect when used with the [async API](RequestBuilder::send_async).
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<String>>) }
    ///
    /// # fn update(caps: &Capabilities) {
    /// caps.http
    ///     .post("https://httpbin.org/json")
    ///     .expect_string()
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn expect_string(self) -> RequestBuilder<Effect, Event, String> {
        let expectation = Box::<ExpectString>::default();
        RequestBuilder {
            req: self.req,
            effect: PhantomData,
            event: PhantomData,
            expectation,
        }
    }

    /// Decode a `T` from a JSON response body prior to dispatching it to the apps `update`
    /// function.
    ///
    /// This has no effect when used with the [async API](RequestBuilder::send_async).
    ///
    /// # Examples
    ///
    /// ```no_run
    /// # use serde::{Deserialize, Serialize};
    /// # struct Capabilities { http: crux_http::Http<Event> }
    /// #[derive(Deserialize)]
    /// struct Response {
    ///     slideshow: Slideshow
    /// }
    ///
    /// #[derive(Deserialize)]
    /// struct Slideshow {
    ///     author: String
    /// }
    ///
    /// enum Event { ReceiveResponse(crux_http::Result<crux_http::Response<Slideshow>>) }
    ///
    /// # fn update(caps: &Capabilities) {
    /// caps.http
    ///     .post("https://httpbin.org/json")
    ///     .expect_json::<Slideshow>()
    ///     .send(Event::ReceiveResponse)
    /// # }
    /// ```
    pub fn expect_json<T>(self) -> RequestBuilder<Effect, Event, T>
    where
        T: DeserializeOwned + 'static,
    {
        let expectation = Box::<ExpectJson<T>>::default();
        RequestBuilder {
            req: self.req,
            effect: PhantomData,
            event: PhantomData,
            expectation,
        }
    }
}

impl<Effect, Event> fmt::Debug for RequestBuilder<Effect, Event> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(&self.req, f)
    }
}
