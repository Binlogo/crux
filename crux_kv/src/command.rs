//! The Command based API for crux_kv

use std::{future::Future, marker::PhantomData};

use crux_core::{command::RequestBuilder, Command, Request};

use super::{KeyValueOperation, KeyValueResult};

pub struct KeyValue<Effect, Event> {
    // Allow the impl to declare trait bounds once. Thanks rustc
    effect: PhantomData<Effect>,
    event: PhantomData<Event>,
}

impl<Effect, Event> KeyValue<Effect, Event>
where
    Effect: Send + From<Request<KeyValueOperation>> + 'static,
    Event: Send + 'static,
{
    pub fn get(
        key: impl Into<String>,
    ) -> RequestBuilder<Effect, Event, impl Future<Output = KeyValueResult>> {
        Command::request_from_shell(KeyValueOperation::Get { key: key.into() })
    }

    pub fn set(
        key: impl Into<String>,
        value: Vec<u8>,
    ) -> RequestBuilder<Effect, Event, impl Future<Output = KeyValueResult>> {
        Command::request_from_shell(KeyValueOperation::Set {
            key: key.into(),
            value,
        })
    }

    pub fn delete(
        key: impl Into<String>,
    ) -> RequestBuilder<Effect, Event, impl Future<Output = KeyValueResult>> {
        Command::request_from_shell(KeyValueOperation::Delete { key: key.into() })
    }

    pub fn exists(
        key: impl Into<String>,
    ) -> RequestBuilder<Effect, Event, impl Future<Output = KeyValueResult>> {
        Command::request_from_shell(KeyValueOperation::Exists { key: key.into() })
    }

    pub fn list_keys(
        prefix: impl Into<String>,
        cursor: u64,
    ) -> RequestBuilder<Effect, Event, impl Future<Output = KeyValueResult>> {
        Command::request_from_shell(KeyValueOperation::ListKeys {
            prefix: prefix.into(),
            cursor,
        })
    }
}
