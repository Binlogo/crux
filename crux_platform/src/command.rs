use std::{future::Future, marker::PhantomData};

use crux_core::{command::RequestBuilder, Command, Request};

use crate::{PlatformRequest, PlatformResponse};

pub struct Platform<Effect, Event> {
    effect: PhantomData<Effect>,
    event: PhantomData<Event>,
}

impl<Effect, Event> Platform<Effect, Event>
where
    Effect: From<Request<PlatformRequest>> + Send + 'static,
    Event: Send + 'static,
{
    /// Get the platform of the shell
    #[must_use]
    pub fn get() -> RequestBuilder<Effect, Event, impl Future<Output = PlatformResponse>> {
        Command::request_from_shell(PlatformRequest)
    }
}
