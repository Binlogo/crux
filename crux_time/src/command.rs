use std::{
    future::Future,
    marker::PhantomData,
    time::{Duration, SystemTime},
};

use crux_core::{command::RequestBuilder, Command, Request};
use futures::{
    channel::oneshot::{self, Sender},
    select, FutureExt,
};

use crate::{get_timer_id, TimeRequest, TimeResponse, TimerId};

pub struct Time<Effect, Event> {
    // Allow impl level trait bounds to avoid repetition
    effect: PhantomData<Effect>,
    event: PhantomData<Event>,
}

impl<Effect, Event> Time<Effect, Event>
where
    Effect: Send + From<Request<TimeRequest>> + 'static,
    Event: Send + 'static,
{
    /// Ask for the current wall-clock time.
    pub fn now() -> RequestBuilder<Effect, Event, impl Future<Output = TimeResponse>> {
        Command::request_from_shell(TimeRequest::Now)
    }

    /// Ask to receive a notification when the specified
    /// [`SystemTime`](std::time::SystemTime) has arrived.
    pub fn notify_at(
        system_time: SystemTime,
    ) -> (
        RequestBuilder<Effect, Event, impl Future<Output = TimeResponse>>,
        TimerHandle,
    ) {
        let id = get_timer_id();
        let (sender, mut receiver) = oneshot::channel();

        let builder = RequestBuilder::new(move |ctx| async move {
            select! {
                response = ctx.request_from_shell(
                    TimeRequest::NotifyAt {
                        id,
                        instant: system_time.into()
                    }
                ).fuse() =>  response,
                cleared = receiver => {
                    // The Err variant would mean the sender was dropped,
                    // but `receiver` is a fused future,
                    // which signals `is_terminated` true in that case,
                    // so this branch of the select will
                    // never run for the Err case

                    let id = cleared.unwrap();
                    ctx.request_from_shell(TimeRequest::Clear { id }).await
                }
            }
        });

        let handle = TimerHandle {
            timer_id: id,
            abort: sender,
        };

        (builder, handle)
    }

    /// Ask to receive a notification after the specified
    /// [`Duration`](std::time::Duration) has elapsed.
    pub fn notify_after(
        duration: Duration,
    ) -> (
        RequestBuilder<Effect, Event, impl Future<Output = TimeResponse>>,
        TimerHandle,
    ) {
        let id = get_timer_id();
        let (sender, mut receiver) = oneshot::channel();

        let builder = RequestBuilder::new(move |ctx| async move {
            select! {
                response = ctx.request_from_shell(
                    TimeRequest::NotifyAfter {
                        id,
                        duration: duration.into()
                    }
                ).fuse() => response,
                cleared = receiver => {
                    // The Err variant would mean the sender was dropped,
                    // but `receiver` is a fused future,
                    // which signals `is_terminated` true in that case,
                    // so this branch of the select will
                    // never run for the Err case
                    let id = cleared.unwrap();
                    ctx.request_from_shell(TimeRequest::Clear { id }).await
                }
            }
        });

        let handle = TimerHandle {
            timer_id: id,
            abort: sender,
        };

        (builder, handle)
    }
}

/// A handle to a requested timer. Allows the timer to be cleared. The handle is safe to drop,
/// in which case the original timer is no longer abortable
#[derive(Debug)]
pub struct TimerHandle {
    timer_id: TimerId,
    abort: Sender<TimerId>,
}

impl Eq for TimerHandle {}

impl PartialEq for TimerHandle {
    fn eq(&self, other: &Self) -> bool {
        self.timer_id == other.timer_id
    }
}

impl PartialEq<TimerId> for TimerHandle {
    fn eq(&self, other: &TimerId) -> bool {
        self.timer_id == *other
    }
}

impl PartialEq<TimerHandle> for TimerId {
    fn eq(&self, other: &TimerHandle) -> bool {
        *self == other.timer_id
    }
}

impl TimerHandle {
    /// Clear the associated timer request.
    /// The shell will be notified that the timer has been cleared
    /// with `TimeRequest::Clear { id }`,
    /// so it can clean up associated resources.
    /// The original task will resolve
    /// with `TimeResponse::Cleared { id }`.
    pub fn clear(self) {
        let _ = self.abort.send(self.timer_id);
    }
}

#[cfg(test)]
mod tests {
    use std::time::Duration;

    use crux_core::Request;

    use super::Time;
    use crate::{TimeRequest, TimeResponse};

    enum Effect {
        Time(Request<TimeRequest>),
    }

    impl From<Request<TimeRequest>> for Effect {
        fn from(value: Request<TimeRequest>) -> Self {
            Self::Time(value)
        }
    }

    #[derive(Debug, PartialEq)]
    enum Event {
        Elapsed(TimeResponse),
    }

    #[test]
    fn timer_can_be_cleared() {
        let (cmd, handle) = Time::notify_after(Duration::from_secs(2));
        let mut cmd = cmd.then_send(Event::Elapsed);

        let effect = cmd.effects().next();

        assert!(cmd.events().next().is_none());

        let Some(Effect::Time(_request)) = effect else {
            panic!("should get an effect");
        };

        handle.clear();

        let effect = cmd.effects().next();
        assert!(cmd.events().next().is_none());

        let Some(Effect::Time(mut request)) = effect else {
            panic!("should get an effect");
        };

        let TimeRequest::Clear { id } = request.operation else {
            panic!("expected a Clear request");
        };

        request
            .resolve(TimeResponse::Cleared { id })
            .expect("should resolve");

        let event = cmd.events().next();

        assert!(matches!(
            event,
            Some(Event::Elapsed(TimeResponse::Cleared { .. }))
        ));
    }

    #[test]
    fn dropping_a_timer_handle_does_not_clear_the_request() {
        let (cmd, handle) = Time::notify_after(Duration::from_secs(2));
        drop(handle);

        let mut cmd = cmd.then_send(Event::Elapsed);
        let effect = cmd.effects().next();

        assert!(cmd.events().next().is_none());

        let Some(Effect::Time(mut request)) = effect else {
            panic!("should get an effect");
        };

        let TimeRequest::NotifyAfter { id, .. } = request.operation else {
            panic!("Expected a NotifyAfter");
        };

        request
            .resolve(TimeResponse::DurationElapsed { id })
            .expect("should resolve");

        let event = cmd.events().next();

        assert_eq!(
            event,
            Some(Event::Elapsed(TimeResponse::DurationElapsed { id }))
        );
    }
}
