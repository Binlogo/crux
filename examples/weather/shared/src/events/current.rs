use crux_core::{render::render, Command};
use crux_http::command::Http;
use serde::{Deserialize, Serialize};

use crate::workflows::home::HomeEvent;
use crate::{CurrentResponse, Effect, Event, Model};

const WEATHER_URL: &str = "https://api.openweathermap.org/data/2.5/weather";
const API_KEY: &str = "42005d273a8a49c88a8173878232508";

#[derive(Serialize)]
pub struct CurrentQueryString {
    pub lat: String,
    pub lon: String,
    pub appid: &'static str,
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub enum CurrentWeatherEvent {
    #[serde(skip)]
    Fetch(f64, f64),
    #[serde(skip)]
    SetWeather(crux_http::Result<crux_http::Response<CurrentResponse>>),
}

pub fn update(event: CurrentWeatherEvent, model: &mut Model) -> Command<Effect, Event> {
    match event {
        CurrentWeatherEvent::Fetch(lat, long) => Http::get(WEATHER_URL)
            .expect_json()
            .query(&CurrentQueryString {
                lat: lat.to_string(),
                lon: long.to_string(),
                appid: API_KEY,
            })
            .expect("could not serialize query string")
            .build()
            .then_send(|result| Event::CurrentWeather(CurrentWeatherEvent::SetWeather(result))),
        CurrentWeatherEvent::SetWeather(result) => {
            model.weather_data = result.unwrap().take_body().unwrap();
            render()
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crux_core::App as _;
    use crux_http::{
        protocol::{HttpRequest, HttpResponse, HttpResult},
        testing::ResponseBuilder,
    };

    #[test]
    fn test_current_weather_fetch() {
        let lat_lng = (33.456789, -112.037222);
        let event = CurrentWeatherEvent::Fetch(lat_lng.0, lat_lng.1);

        let mut cmd = update(event, &mut Model::default());

        let mut request = cmd.effects().next().unwrap().expect_http();

        assert_eq!(
            &request.operation,
            &HttpRequest::get(WEATHER_URL)
                .query(&CurrentQueryString {
                    lat: lat_lng.0.to_string(),
                    lon: lat_lng.1.to_string(),
                    appid: API_KEY,
                })
                .expect("could not serialize query string")
                .build()
        );

        // Test response handling
        request
            .resolve(HttpResult::Ok(
                HttpResponse::ok()
                    .body(
                        r#"{
                            "main": {
                                "temp": 20.0,
                                "feels_like": 18.0,
                                "temp_min": 18.0,
                                "temp_max": 22.0,
                                "pressure": 1013,
                                "humidity": 50
                            },
                            "coord": {
                                "lat": 33.456789,
                                "lon": -112.037222
                            },
                            "weather": [{
                                "id": 800,
                                "main": "Clear",
                                "description": "clear sky",
                                "icon": "01d"
                            }],
                            "base": "",
                            "visibility": 10000,
                            "wind": {
                                "speed": 4.1,
                                "deg": 280,
                                "gust": 5.2
                            },
                            "clouds": {
                                "all": 0
                            },
                            "dt": 1716216000,
                            "sys": {
                                "id": 1,
                                "country": "US",
                                "type": 1,
                                "sunrise": 1716216000,
                                "sunset": 1716216000
                            },
                            "timezone": 1,
                            "id": 1,
                            "name": "Phoenix",
                            "cod": 200
                        }"#,
                    )
                    .build(),
            ))
            .unwrap();

        let actual = cmd.events().next().unwrap();
        assert!(matches!(
            actual,
            Event::CurrentWeather(CurrentWeatherEvent::SetWeather(_))
        ));
    }
}
