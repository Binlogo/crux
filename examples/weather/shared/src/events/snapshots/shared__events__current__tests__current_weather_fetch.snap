---
source: shared/src/events/current.rs
expression: model.weather_data
---
coord:
  lat: 33.456789
  lon: -112.037222
weather:
  - id: 800
    main: Clear
    description: clear sky
    icon: 01d
base: ""
main:
  temp: 20
  feels_like: 18
  temp_min: 18
  temp_max: 22
  pressure: 1013
  humidity: 50
visibility: 10000
wind:
  speed: 4.1
  deg: 280
  gust: 5.2
clouds:
  all: 0
dt: 1716216000
sys:
  type: 1
  id: 1
  country: US
  sunrise: 1716216000
  sunset: 1716216000
timezone: 1
id: 1
name: Phoenix
cod: 200
